experimental/dds/tree2/src/feature-libraries/object-forest/objectForest.ts

```diff
@@ -141,41 +127,74 @@ class ObjectForest extends SimpleDependee implements IEditableForest {
+				this.detachEdit(range, undefined);
 			},
-			onInsert(index: number, content: Delta.ProtoNode[]): void {
-				const range = this.forest.add(content);
-				moveIn(index, range, this.cursor);
+			create(index: PlaceIndex, content: Delta.ProtoNodes): void {
+				this.forest.invalidateDependents();
+				const nodes = Array.from(content, mapTreeFromCursor);
+				const [parent, key] = this.cursor.getParent();
+				const destinationField = getMapTreeField(parent, key, true);
+				assertValidIndex(index, destinationField, true);
+				// TODO: this will fail for very large insertions due to argument limits.
+				destinationField.splice(index, 0, ...nodes);
+			},
+			attach(source: DetachedRangeUpPath, destination: PlaceIndex): void {
+				this.forest.invalidateDependents();
+				this.attachEdit(source, destination);
 			},
-			onMoveOut(index: number, count: number, id?: Delta.MoveId): void {
+			detach(source: Range, destination: DetachedPlaceUpPath): void {
 				this.forest.invalidateDependents();
+				this.detachEdit(source, destination);
+			},
+			attachEdit(source: DetachedRangeUpPath, destination: PlaceIndex): void {
 				const [parent, key] = cursor.getParent();
-				const sourceField = getMapTreeField(parent, key, false);
-				const startIndex = index;
-				const endIndex = index + count;
-				assertValidIndex(startIndex, sourceField, true);
-				assertValidIndex(endIndex, sourceField, true);
-				assert(
-					startIndex <= endIndex,
-					0x371 /* detached range's end must be after its start */,
-				);
-				const newField = sourceField.splice(startIndex, endIndex - startIndex);
-				const field = this.forest.addFieldAsDetached(newField);
-				if (id !== undefined) {
-					moves.set(id, field);
-				} else {
-					this.forest.delete(field);
-				}
+				const currentField = getMapTreeField(parent, key, true);
+				assertValidIndex(destination, currentField, true);
+				const sourceField = getMapTreeField(this.forest.roots, source.field, false);
+				assertValidIndex(source.start, sourceField, true);
+				assertValidIndex(source.end, sourceField, true);
+				assert(source.end >= source.start, "Range end must be >= its start");
+				const content = sourceField.splice(source.start, source.end - source.start);
 				if (sourceField.length === 0) {
+					this.forest.roots.fields.delete(source.field);
+				}
+				// TODO: this will fail for very large insertions due to argument limits.
+				currentField.splice(destination, 0, ...content);
+			},
+			detachEdit(source: Range, destination: DetachedPlaceUpPath | undefined): void {
+				const [parent, key] = cursor.getParent();
+				const currentField = getMapTreeField(parent, key, true);
+				assertValidIndex(source.start, currentField, true);
+				assertValidIndex(source.end, currentField, true);
+				assert(source.end >= source.start, "Range end must be >= its start");
+				const content = currentField.splice(source.start, source.end - source.start);
+				if (currentField.length === 0) {
```

Comment:
I think its safer to do the empty case cleanup at the end, since that avoids messing up cases where src and dst are the same.